python/hsfs/core/arrow_flight_client.py

```diff
@@ -46,130 +54,253 @@ def close():
+    WARNING_FULL_QUEUE_ERROR = "No free slots available. "
+    WARNING_TIMEOUT_ERROR = "Request timed out. "
+    DEFAULT_TIMEOUT_SECONDS = 900
+    DEFAULT_HEALTHCHECK_TIMEOUT_SECONDS = 5
+    DEFAULT_GRPC_MIN_RECONNECT_BACKOFF_MS = 2000
 
     def __init__(self):
+        _logger.debug("Initializing Hopsworks Feature Query Service Client.")
+        self._timeout: float = self.DEFAULT_TIMEOUT_SECONDS
+        self._health_check_timeout: float = self.DEFAULT_HEALTHCHECK_TIMEOUT_SECONDS
+
+        self._enabled_on_cluster: bool = False
+        self._disabled_for_session: bool = False
+        self._host_url: Optional[str] = None
+        self._connection: Optional[pyarrow.flight.FlightClient] = None
+
+        self._client = client.get_instance()
+        self._variable_api: VariableApi = VariableApi()
+
         try:
-            self._variable_api = VariableApi()
-            self._is_enabled = self._variable_api.get_flyingduck_enabled()
-        except Exception:
-            # if feature flag cannot be retrieved, assume it is disabled
-            self._is_enabled = False
+            self._check_cluster_service_enabled()
+            self._host_url = self._retrieve_host_url()
+
+            if self._enabled_on_cluster:
+                _logger.debug(
+                    "Hopsworks Feature Query Service is enabled on the cluster."
+                )
+                self._initialize_flight_client()
+            else:
+                _logger.debug(
+                    "Hopsworks Feature Query Service Client is not enabled on the cluster or a cluster variable is misconfigured."
+                )
+                self._disable_for_session()
+                return
+        except Exception as e:
+            _logger.debug("Failed to connect to Hopsworks Feature Query Service")
+            _logger.exception(e)
+            self._disable_for_session(str(e))
             return
 
-        if self._is_enabled:
-            try:
-                self._initialize_connection()
-            except Exception as e:
-                self._disable(str(e))
-
-    def _disable(self, message):
-        self._is_enabled = False
-        warnings.warn(
-            f"Could not establish connection to ArrowFlight Server. ({message}) "
-            f"Will fall back to hive/spark for this session. "
-            f"If the error persists, you can disable using ArrowFlight "
-            f"by changing the cluster configuration (set 'enable_flyingduck'='false')."
-        )
+        try:
+            self._health_check()
+            self._register_certificates()
+        except Exception as e:
+            _logger.debug("Failed to connect to Hopsworks Feature Query Service")
+            _logger.exception(e)
+            warnings.warn(
+                f"Failed to connect to Hopsworks Feature Query Service, got {str(e)}."
+                + self.DEFAULTING_TO_DIFFERENT_SERVICE_WARNING
+                + self.CLIENT_WILL_STAY_ACTIVE_WARNING
+            )
```

Comment:
need to call `_disable_for_session` here?

Reply:
No it is what the message is saying, the client will stay active and they can explicitly override the behaviour.